glommio/src/io/dma_file_stream.rs

```diff
@@ -904,56 +907,72 @@ impl DmaStreamWriterState {
         self.flushed_pos
     }
 
-    fn adjust_flushed_pos(&mut self, just_written: u64) {
-        self.out_of_order_write_returns.push(just_written);
-        self.out_of_order_write_returns.sort_by_key(|x| Reverse(*x));
-        while let Some(x) = self.out_of_order_write_returns.last() {
-            if *x == self.flushed_pos {
-                self.flushed_pos += self.buffer_size as u64;
-                self.out_of_order_write_returns.pop();
-            } else {
-                return;
+    fn on_flush_success(&mut self, flush_pos: u64) {
+        let mut complete_prefix_len = 0usize;
+        let mut complete_prefix_finalized = false;
+        let mut found_index = None;
+        for (i, (pos, status)) in self.flushes.iter_mut().enumerate() {
```

Comment:
how is this easier to follow than a map?

Reply:
Fair point! It's not just the `pending` map though. There is also the `out_of_order_writes` vec. The `flushes` ringbuffer combines the two. I found it easier to reason about the prefix of contiguous completions in this ring buffer representing a safe update to `state.flushed_pos`. Now there is a single place assignment to `state.flushed_pos` happens.

Reply:
Also not sure how I can make non-aligned positions work with the current approach taken in `adjust_flushed_pos()`.